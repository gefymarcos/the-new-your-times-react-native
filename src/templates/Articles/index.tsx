/* eslint-disable @typescript-eslint/no-explicit-any */
import * as React from 'react'
import {
  ScrollView,
  StyleSheet,
  ScrollViewProps,
  ActivityIndicator,
  Text
} from 'react-native'
import { useScrollToTop } from '@react-navigation/native'
import { loadArticlesAction } from '../../modules/articles/actions'
import { useDispatch, useSelector } from 'react-redux'
import { ArticleProps } from 'modules/articles/types'
import BackDrop from '../../components/Modal'
import Card from '../../components/Card'
import Container from '../../components/Container'
import theme from '../../styles/theme'

type articleTemplateProps = {
  section?: 'science' | 'technology'
} & Partial<ScrollViewProps>

export default function Articles({
  section = 'science',
  ...rest
}: articleTemplateProps) {
  const dispatch = useDispatch()
  const articles = useSelector((state: any) => state?.articles?.data)
  const error = useSelector((state: any) => state?.articles?.errorMsg)
  const article = useSelector((state: any) => state?.articles?.articleDetail)

  React.useEffect(() => {
    dispatch(loadArticlesAction(section))
  }, [dispatch, section])

  const ref = React.useRef<ScrollView>(null)

  useScrollToTop(ref)

  return (
    <>
      {!!article?.title && <BackDrop {...article} />}
      {!!error && (
        <Container alignContent="center">
          <Text>
            Não foi possível carregar as informações agora, tente novamente mais
            tarde.
          </Text>
        </Container>
      )}
      {articles.length === 0 && !error && !article?.title && (
        <Container alignContent="center">
          <ActivityIndicator size="large" color={theme.colors.secondary} />
        </Container>
      )}
      {!!articles && (
        <ScrollView
          ref={ref}
          style={styles.container}
          contentContainerStyle={styles.content}
          {...rest}
        >
          {articles.map((article: ArticleProps, index: number) => (
            <Card
              key={index}
              title={article.title}
              abstract={article.abstract}
              url={article.url}
            />
          ))}
        </ScrollView>
      )}
    </>
  )
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: theme.colors.primary
  },
  content: {
    paddingVertical: theme.spacings.xsmall
  }
})
