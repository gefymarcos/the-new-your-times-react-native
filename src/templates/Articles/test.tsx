import React from 'react'
import { render } from '@testing-library/react-native'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import Articles from '.'

jest.useFakeTimers()

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn().mockReturnValue([
    {
      title: 'teste',
      abstract: 'teste',
      url: 'teste'
    }
  ])
}))

describe('<Articles />', () => {
  const middlewares = [thunk]
  const mockStore = configureStore(middlewares)
  const store = mockStore()

  it('should render correctly the home with tabs', () => {
    const { getByTestId, getAllByText } = render(
      <Provider store={store}>
        <Articles />
      </Provider>
    )

    const scrollView = getByTestId(/scrollView/i)
    const title = getAllByText(/teste/i)

    expect(scrollView).toBeDefined()
    expect(title).toBeDefined()
  })

  it('should render correctly based on snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={mockStore({})}>
          <Articles />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
