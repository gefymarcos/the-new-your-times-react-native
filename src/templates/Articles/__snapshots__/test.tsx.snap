// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Articles /> should render correctly based on snapshot 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "paddingTop": 20,
        },
        Object {
          "alignItems": "center",
          "justifyContent": "flex-end",
        },
      ]
    }
  >
    <Text>
      Não foi possível carregar as informações agora, tente novamente mais tarde.
    </Text>
  </View>,
  <RCTScrollView
    contentContainerStyle={
      Object {
        "paddingVertical": 16,
      }
    }
    style={
      Object {
        "backgroundColor": "#FAFAFA",
      }
    }
    testID="scrollView"
  >
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderBottomColor": "#212121",
            "borderBottomWidth": 1,
            "marginHorizontal": 16,
          }
        }
      >
        <View
          style={
            Object {
              "justifyContent": "center",
              "paddingVertical": 32,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#212121",
                "fontSize": 18,
                "fontWeight": "600",
                "marginHorizontal": 16,
                "marginVertical": 8,
              }
            }
          >
            teste
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`
